transformations/transformer_replace/README.md

```diff
@@ -0,0 +1,33 @@
+# Transformer Replace
+
+This perturbation replaces words based on recommendations from a Huggingface transformer model.
+
+Author name: Gautier Dagan
+Author email: **
+Author Affiliation: **
+
+## What type of a transformation is this?
+
+This transformation acts as a way to augment the text in way to change certain words according to some which are most likely under a transformer model.
+
+## What tasks does it intend to benefit?
+
+This perturbation would benefit all tasks which have a sentence/paragraph/document as input like text classification,
+text generation, etc.
+
+`python evaluate.py -t TransformerReplace -task TEXT_CLASSIFICATION`
+`model_name = "aychang/roberta-base-imdb"`
+
+The accuracy of a RoBERTa model (fine-tuned on IMDB) (model: "aychang/roberta-base-imdb")
+on a subset of IMDB sentiment dataset = 95.74
+The accuracy of the same model on the perturbed set = 88.26
+
+The average bleu score of a distillbert model (fine-tuned on xsum) (model: "sshleifer/distilbart-xsum-12-6")
+on a subset (10%) of xsum test dataset = 14.9104
+The average bleu score of same model on the pertubed set = 11.9221
+
+## Previous Work
```

Comment:
Some work that I know use LMs to do contextual word replacement -- I'd suggest consider adding refs to them!

[Beyond Accuracy: Behavioral Testing of NLP Models with CheckList](https://homes.cs.washington.edu/~marcotcr/acl20_checklist.pdf)

[Conditional BERT Contextual Augmentation](https://arxiv.org/abs/1812.06705)

[Contextualized Perturbation for Textual Adversarial Attack](https://arxiv.org/pdf/2009.07502.pdf)


I'm also interested in what the author see as the difference between this pull request and https://github.com/GEM-benchmark/NL-Augmenter/pull/81 ? Seems the two share the same spirit as far as I would concern?

Reply:
Hey @tongshuangwu, thank you for the review 🙌 ! 
I can certainly cite these papers - I did cite a few but it seems that you reviewed the first commit on my branch rather than the last? I already cite `Conditional BERT Contextual Augmentation`, but will add the rest.

Definitely a good point as well, when I drafted this PR originally I distinctly remember that #81 was doing things differently, but it seems now that there is indeed a lot overlap with it if you consider the base defaults.

However this approach does **two** main things differently:
1. It offers a way to select the words to replace - through the use of a spacy model and based on POS tokens. Because it uses Spacy it is probably also suitable to use for other languages as long as the spacy model and the huggingface are targeting the same underlying language.

2. It offers a way to _prime_ the model and context. A big problem in these types of augmentation is that meaning can be changed. Giving a way to do so by prepending some context to the sentence to augment can be a way to limit the meaning to drift. 
For instance if I was doing sentiment analysis and have the sentence "We had some delicious food at this restaurant", and that I am replacing every adjective with a suggestion from the model, then I would mask delicious to have: "We had some MASK food at this restaurant". A model could then potentially complete the MASK with "terrible" or "aweful" and completely change the meaning of the sentence. 
However if I prime the context with for instance "POSITIVE : " so that my masked sentence becomes "POSITIVE : We had some MASK food at this restaurant", the model will be less likely to change the meaning of the sentence and thus give an augmentation that might correspond better to the target.

Both of these are in fact points that @kaustubhdhole makes on #81 (https://github.com/GEM-benchmark/NL-Augmenter/pull/81#issuecomment-880671719
) 😄 

This all means that the randomness of the augmentation can be somewhat better controlled through the word selection mechanism and through additional context.